File: apps/divsoup/lib/divsoup/analyzer/achievements/div_soup/silver.ex
defmodule Divsoup.Achievement.DivSoupSilver do
  alias Divsoup.Achievement

  @behaviour Divsoup.AchievementRule

  @impl true
  def evaluate(html_tree, _) do
    ratio = Divsoup.Util.DivSoup.get_div_ratio(html_tree)

    if ratio > 0.50 do
      []
    else
      ["Only #{ratio * 100}% of the HTML elements in the page are divs"]
    end
  end

  @impl true
  def achievement() do
    %Achievement{
      hierarchy: :silver,
      title: "Div Soup",
      group: "div_soup",
      description: "More than <strong>50%</strong> of the HTML elements in the page are <code>&lt;div&gt;</code> elements"
    }
  end
end

File: src/achievements/div_soup/silver.ts
import type { AchievementRule } from "../types";
import { getDivRatio } from "../utils";

export const rule: AchievementRule = {
  id: "div_soup.silver",
  title: "Div Soup",
  group: "div_soup",
  description:
    "More than <strong>50%</strong> of the HTML elements in the page are <code>&lt;div&gt;</code> elements",
  hierarchy: "silver",
  evaluate: ({ doc }) => getDivRatio(doc) > 0.5,
};
